### PR TITLE
scheduler performance improvement for get_svr_index

### PR DESCRIPTION
**Pull request by suresh-thelkar**
#### Describe Bug or Feature

As of now get_svr_index() is being used inside Scheduler code heavily especially under the following scenarios.

- For every job 
- For every node

This happens for every cycle when we query the data from Server/s. get_svr_index() loops through all svr_conns given a server name to find out the index associated with that server. Looping svr_conns array for every job and node is very costly and is O(n) time complexity where n is proportional to number of servers. The function also does a strcmp to match the server name with the given name which is again a costly operation. This enhancement is an attempt to not use get_svr_index() in the above situations and makes Scheduler to get server index in O(1) time in the above scenarios. 

In fact Scheduler now uses get_svr_index() only once when Server/s are connecting to Scheduler at the time of connection initiation which is required anyways.

Also removed some unneeded code from Scheduler as part of this work.

Just out of curiosity, I have run the following performance test in a performance machine before and after this enhancement. Following are the results.  Have selected this scenario purposefully just to increase load on Scheduler. 
Scenario: Make Scheduling off and Submit 50,000 jobs, then enable scheduling and measure the performance.

  | Servers | Moms | ncpus | Jobs | Sched   Throughput     in sec | End to end   throughput     in sec | jobs/sec
-- | -- | -- | -- | -- | -- | -- | --
Integration   branch | 10 | 200 | 250 | 50000 | 9.03 | 16 | 3125
PR branch | 10 | 200 | 250 | 50000 | 8.01 | 15 | 3333.333333
  |   |   |   |   |   | Increase in jobs/sec | 208.3333333

By looking at the above numbers it looks like it is definitely increasing performance. We can get better results if we increase the load on Scheduler heavily. 50,000 jobs also to Scheduler seems less to verify the said scenario. May be we can try with more servers and thousands of moms in which case the difference would be more substantial.

### CODE REVIEW

**Comment by suresh-thelkar** [comment]
@subhasisb, As you are aware of the approach followed, I request you to review this and provide your feedback.

**Comment by agrawalravi90** [comment]
@suresh-thelkar can you please try with more jobs and nodes? the performance improvement doesn't seem considerable with 50k jobs. You don't really need thousands of moms, you don't even need to run any jobs. If you have 1 million jobs, just 1 node, and run 1 sched cycle, you should see the performance effect of this change as the scheduler will still find the server index associated with all jobs before dismissing all of them

**Comment by agrawalravi90** [comment]
Also, can you please explain your change? what exactly did you do to make things faster?

File: src/lib/Libutil/misc_utils.c [hunk 1]
```diff
@@ -2268,3 +2268,25 @@ rand_num()
...
+ * @retval -1 for failure
+ */
+int
+get_svr_index_sock(int sock, svr_conn_t *svr_conns)
+{
+	int i;
+	int svrindex = -1;
+
+	for (i = 0; i < get_num_servers(); i++) {
+		if (svr_conns[i].sd == sock) {
```
Comment (agrawalravi90) [inline src/lib/Libutil/misc_utils.c:2285]: won't 'sock' always point to the fd of just one server? I don't think this will work

Comment (agrawalravi90) [reply src/lib/Libutil/misc_utils.c:2285]: ok no, it will work, we loop over sd(s) of all servers in __pbs_selstat .. but you won't really need to do this, you can just pass the server index from __pbs_selstat to PBSD_status_get() right?

File: src/lib/Libifl/int_status.c [hunk 2]
```diff
@@ -574,6 +581,36 @@ PBSD_status_get(int c)
...
+				/* Memory of 3 bytes because we at most can have 99 servers.  So to represent this in string
+				we need an array of len 2 for server index + 1 for NULL char */
+				pat->value = malloc(3);
+				if (pat->value == NULL) {
+					pbs_errno = PBSE_SYSTEM;
+					free_attrl(pat);
+					free(pat->name);
+					return NULL;
+				}
+				sprintf(pat->value, "%d", get_svr_index_sock(c, svr_conns));
```
[the remark this answers is not in the document]
Comment (agrawalravi90) [reply src/lib/Libifl/int_status.c:608]: I think you can remove the call to get_svr_index_sock, you already know the server index inside __pbs_selstat, just pass it to PBSD_status_get .. you don't even need 'type' argument, if index > 0 then add it

[the remark this answers is not in the document]
Comment (suresh-thelkar) [reply src/lib/Libifl/int_status.c:608]: > It seems like we are still going to go loop over the list of all servers for every job and node, every scheduling cycle? How is this O(1) over O(N) ?

There is no new loop introduced. We are not looping over the list of servers for every job and node. All we are doing is 

- Whenever a stat call happens we just add one more additional attribute named "server_index" to the list of attributes that is returned. This is just prepending an attribute to the result of attributes of a stat call.
- While getting a server index associated with a sock we call get_svr_index() with the sock where we loop through all servers to find out the right index. Let's say we have 3 servers this function is called at the most three times. 
- After the above (for example in case of  pbs_statvnode()), all the nodes will automatically have the server_index attribute. So Scheduler does not need to call find_svr_index() for each and every node.
- So basically once the Scheduler gets a reply for a stat call it access server index in O(1) time. It is because it does not need to call the function get_svr_index() for every node/job.

[the remark this answers is not in the document]
Comment (suresh-thelkar) [reply src/lib/Libifl/int_status.c:608]: > I think you can remove the call to get_svr_index_sock, you already know the server index inside __pbs_selstat, just pass it to PBSD_status_get .. you don't even need 'type' argument, if index > 0 then add it

At places we will not get server index for example in PBSD_status(). But we can workout this by sending -1 as server_index to this function. We actually use sever index only in case of the following batch requests. Yes we can eliminate this function get_svr_index_sock() and doing the changes accordingly. 
type is definitely needed because I want to add server_index attribute for only the following batch requests. I will push my changes once my testing is done and add as separate commit.

- PBS_BATCH_StatusJob 
- PBS_BATCH_SelStat 
- PBS_BATCH_StatusNode)

File: src/lib/Libutil/misc_utils.c (hunk 1 above)
Comment (suresh-thelkar) [reply src/lib/Libutil/misc_utils.c:2285]: As per the modified code this function is not needed. I am removing it.

**Comment by suresh-thelkar** [comment]
> Also, can you please explain your change? what exactly did you do to make things faster?

I have explained the changes here https://github.com/subhasisb/openpbs/pull/122#discussion_r465580331

**Comment by suresh-thelkar** [comment]
> @suresh-thelkar can you please try with more jobs and nodes? the performance improvement doesn't seem considerable with 50k jobs. You don't really need thousands of moms, you don't even need to run any jobs. If you have 1 million jobs, just 1 node, and run 1 sched cycle, you should see the performance effect of this change as the scheduler will still find the server index associated with all jobs before dismissing all of them

I will run the performance test after all the review comments. Now that I have removed the function get_svr_index_sock() also, theoretically accessing server index became O(1) from O(n). However it is always good to test this.

**Comment by suresh-thelkar** [comment]
@agrawalravi90, I have implemented your comments. Please take a look and give further feedback.

File: src/lib/Libifl/int_status.c [hunk 3]
```diff
@@ -574,6 +574,37 @@ PBSD_status_get(int c)
 				stp->brp_attrl = 0;
 			bsp->next = NULL;
 			rbsp->last = bsp;
+
+			if (type == PBS_BATCH_StatusJob || type == PBS_BATCH_SelStat || type == PBS_BATCH_StatusNode) {
```
Comment (agrawalravi90) [inline src/lib/Libifl/int_status.c:578]: if it's only for the scheduler then you don't need to check for PBS_BATCH_StatusJob right?

File: src/lib/Libifl/int_status.c [hunk 4]
```diff
@@ -574,6 +574,37 @@ PBSD_status_get(int c)
...
+				we need an array of len 2 for server index + 1 for NULL char */
+				pat->value = malloc(3);
+				if (pat->value == NULL) {
+					pbs_errno = PBSE_SYSTEM;
+					free_attrl(pat);
+					free(pat->name);
+					return NULL;
+				}
+				
+				sprintf(pat->value, "%d", svr_index);
```
Comment (agrawalravi90) [inline src/lib/Libifl/int_status.c:602]: use snprintf

File: src/scheduler/pbs_sched.c [hunk 5]
```diff
@@ -1474,13 +1473,6 @@ socket_to_conn(int sock, struct sockaddr_in saddr_in)
 		return -1;
 
 	if (conn_arr[svr_conn_index].sd == -1) {
-		if ((phe = gethostbyaddr((char *) &saddr_in.sin_addr, sizeof(saddr_in.sin_addr), AF_INET)) == NULL) {
```
Comment (agrawalravi90) [inline src/scheduler/pbs_sched.c:1477]: why did you remove this? is 'host_name' field not used anywhere? If it isn't then maybe we should remove it from the svr_conn_t structure?

**Comment by suresh-thelkar** [comment]
> I have an overall comment about the approach: why not make server index an "internal" server attribute and set it in the server itself (instead of doing it at the client side)? that way it will only be set once, and not be in the critical path as server can set it when the job is first queued (and update it when it's moved from another server).

One of the design goals for the current Multi-Server approach that we have is "PBS_SERVER_INSTANCES" should not be exposed at the Server."
If we go by the above we cannot create and set this attribute at the Server as the Server itself does not know its own index.

**Comment by suresh-thelkar** [comment]
> Also, IFL is a public interface, so making the change here means that an end user can call statnode and see the server index set, and then be confused because it's not supposed to exist, e.g - even if they query only 1 attribute of a node, they'll now get 2 attributes instead. This can also confuse our front end tools.

I understand your concern. I have made the code changes in such a way that these attributes not exposed to the external clients i.e. like status of jobs(pbsnodes -av) or stat of job using qstat does not have this internal attribute "server_index". Please look into my latest commit for more details. Following is the output of these commands after the latest commit.

**pbsnodes -av does not contain server_index attribute**
[root@spark1 pbspro_rjob_test]# pbsnodes -av
spark1
     Mom = spark1
     Port = 15002
     pbs_version = 20.0.0
     ntype = PBS
     state = free
     pcpus = 4
     resources_available.arch = linux
     resources_available.host = spark1
     resources_available.mem = 7897412kb
     resources_available.ncpus = 4
     resources_available.vnode = spark1
     resources_assigned.accelerator_memory = 0kb
     resources_assigned.hbmem = 0kb
     resources_assigned.mem = 0kb
     resources_assigned.naccelerators = 0
     resources_assigned.ncpus = 0
     resources_assigned.vmem = 0kb
     resv_enable = True
     sharing = default_shared
     last_state_change_time = Thu Aug  6 06:00:35 2020
     last_used_time = Thu Aug  6 06:01:30 2020
     server = spark1

spark1-2
     Mom = spark1
     Port = 28002
     pbs_version = 20.0.0
     ntype = PBS
     state = free
     pcpus = 4
     resources_available.arch = linux
     resources_available.host = spark1
     resources_available.mem = 7897412kb
     resources_available.ncpus = 4
     resources_available.vnode = spark1-2
     resources_assigned.accelerator_memory = 0kb
     resources_assigned.hbmem = 0kb
     resources_assigned.mem = 0kb
     resources_assigned.naccelerators = 0
     resources_assigned.ncpus = 0
     resources_assigned.vmem = 0kb
     resv_enable = True
     sharing = default_shared
     last_state_change_time = Thu Aug  6 06:00:35 2020
     last_used_time = Thu Aug  6 06:01:30 2020
     server = spark1-2

[root@spark1 pbspro_rjob_test]# pbsnodes -av|grep -i index
[root@spark1 pbspro_rjob_test]# 
**job stat does not contain server_index attribute**
[root@spark1 pbspro_rjob_test]# qstat -f 9.spark1-2
Job Id: 9.spark1-2
    Job_Name = STDIN
    Job_Owner = pbsroot@spark1
    resources_used.cpupercent = 0
    resources_used.cput = 00:00:00
    resources_used.mem = 348kb
    resources_used.ncpus = 1
    resources_used.vmem = 4368kb
    resources_used.walltime = 00:00:10
    job_state = R
    queue = workq
    server = spark1
    Checkpoint = u
    ctime = Thu Aug  6 06:05:58 2020
    Error_Path = spark1:/home/pbsroot/STDIN.e9
    exec_host = spark1/3
    exec_vnode = (spark1:ncpus=1)
    Hold_Types = n
    Join_Path = n
    Keep_Files = oed
    Mail_Points = a
    mtime = Thu Aug  6 06:06:06 2020
    Output_Path = spark1:/home/pbsroot/STDIN.o9
    Priority = 0
    qtime = Thu Aug  6 06:05:58 2020
    Rerunable = True
    Resource_List.ncpus = 1
    Resource_List.nodect = 1
    Resource_List.place = pack
    Resource_List.select = 1:ncpus=1
    schedselect = 1:ncpus=1
    stime = Thu Aug  6 06:05:58 2020
    session_id = 2358
    jobdir = /home/pbsroot
    substate = 42
    Variable_List = PBS_O_HOME=/home/pbsroot,PBS_O_LANG=en_US.UTF-8,
	PBS_O_LOGNAME=pbsroot,
	PBS_O_PATH=/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/opt/pbs/bin:
	/home/pbsroot/.local/bin:/home/pbsroot/bin,
	PBS_O_MAIL=/var/spool/mail/pbsroot,PBS_O_SHELL=/bin/bash,
	PBS_O_WORKDIR=/home/pbsroot,PBS_O_SYSTEM=Linux,PBS_O_QUEUE=workq,
	PBS_O_HOST=spark1
    euser = pbsroot
    egroup = tstgrp00
    hashname = 9.spark1-2
    queue_rank = 1596693958592
    queue_type = E
    comment = Job run at Thu Aug 06 at 06:05 on (spark1:ncpus=1)
    etime = Thu Aug  6 06:05:58 2020
    run_count = 1
    Submit_arguments = -koed -- /bin/sleep 100
    executable = <jsdl-hpcpa:Executable>/bin/sleep</jsdl-hpcpa:Executable>
    argument_list = <jsdl-hpcpa:Argument>100</jsdl-hpcpa:Argument>
    project = _pbs_project_default
    run_version = 1
    Submit_Host = spark1

[root@spark1 pbspro_rjob_test]# qstat -f 9.spark1-2|grep -i index
[root@spark1 pbspro_rjob_test]#

File: src/lib/Libifl/int_status.c (hunk 3 above)
Comment (suresh-thelkar) [reply src/lib/Libifl/int_status.c:578]: I have removed type in this function. So we don't need the above if condition for the batch requests.

File: src/scheduler/pbs_sched.c (hunk 5 above)
Comment (suresh-thelkar) [reply src/scheduler/pbs_sched.c:1477]: We are not using host_name anywhere. I have already removed it from svr_conn_t structure.

**Comment by suresh-thelkar** [comment]
@agrawalravi90, I have addressed your review comments. Please take a look and provide your further feedback. Also attached are the latest logs in the Multi-Server setup.
[get_svr_index_pr122.txt](https://github.com/subhasisb/openpbs/files/5034548/get_svr_index_pr122.txt)

**Comment by subhasisb** [comment]
> > I have an overall comment about the approach: why not make server index an "internal" server attribute and set it in the server itself (instead of doing it at the client side)? that way it will only be set once, and not be in the critical path as server can set it when the job is first queued (and update it when it's moved from another server).
> 
> One of the design goals for the current Multi-Server approach that we have is "PBS_SERVER_INSTANCES" should not be exposed at the Server."
> If we go by the above we cannot create and set this attribute at the Server as the Server itself does not know its own index.

I think it is very important that we keep the servers NOT knowing about each other, and make PBS_SERVER_INSTANCES a "client only switch". If absolutely necessary, we will add that to server, but not having the servers know about each other is useful in keeping many complexities away.

**Comment by agrawalravi90** [comment]
@suresh-thelkar changes look good to me

**Comment by agrawalravi90** [comment]
> I suggest we make the srv_index an byte (numeric) variable on the job structure itself (at the scheduler) that would avoid us needing to add yet another strcmp

We already store it as a numeric on the job in scheduler

**Comment by suresh-thelkar** [comment]
> @suresh-thelkar changes look good to me

Thanks for your sign off.

**Comment by suresh-thelkar** [comment]
> > I suggest we make the srv_index an byte (numeric) variable on the job structure itself (at the scheduler) that would avoid us needing to add yet another strcmp
> 
> We already store it as a numeric on the job in scheduler

Yes we already store index as an integer(we can modify it as a bitfield of length 8) in both node_info and job_info structures of Scheduler. I think you mean to not convert the server_index to string in PBSD_status_get so that when we access it in the following places we can directly get it as an integer without again converting to integer using strtol().
```

		else if (!strcmp(attrp->name, ATTR_server_index)) {
			count = strtol(attrp->value, &endp, 10);
			if (*endp == '\0')
				ninfo->svr_index = count;
			else
				ninfo->svr_index = -1;
			if (ninfo->svr_index == -1) {
				free_node_info(ninfo);
				return NULL;
			}
		}
```
and
```
		else if (!strcmp(attrp->name, ATTR_server_index)) {
			count = strtol(attrp->value, &endp, 10);
			if (*endp == '\0')
				resresv->svr_index = count;
			else
				resresv->svr_index = -1;
			if (resresv->svr_index == -1) {
				free_resource_resv(resresv);
				return NULL;
			}
		}
```
If I want to keep index as an integer itself I need to modify the following structure  to contain one more field named numeric_value as follows. This new field is only used for the server_index attribute only. Hope that is fine.
```
struct attrl {
	struct attrl *next;
	char	     *name;
	char	     *resource;
	char	     *value;
        unsigned    numeric_value:8,
	enum batch_op 	 op;	/* not used */
};
```
Please let me know if this what you are suggesting.

**Comment by subhasisb** [comment]
> > > I suggest we make the srv_index an byte (numeric) variable on the job structure itself (at the scheduler) that would avoid us needing to add yet another strcmp
> > 
> > 
> > We already store it as a numeric on the job in scheduler
> 
> Yes we already store index as an integer(we can modify it as a bitfield of length 8) in both node_info and job_info structures of Scheduler. I think you mean to not convert the server_index to string in PBSD_status_get so that when we access it in the following places we can directly get it as an integer without again converting to integer using strtol().
> 
> ```
> 
> 		else if (!strcmp(attrp->name, ATTR_server_index)) {
> 			count = strtol(attrp->value, &endp, 10);
> 			if (*endp == '\0')
> 				ninfo->svr_index = count;
> 			else
> 				ninfo->svr_index = -1;
> 			if (ninfo->svr_index == -1) {
> 				free_node_info(ninfo);
> 				return NULL;
> 			}
> 		}
> ```
> 
> and
> 
> ```
> 		else if (!strcmp(attrp->name, ATTR_server_index)) {
> 			count = strtol(attrp->value, &endp, 10);
> 			if (*endp == '\0')
> 				resresv->svr_index = count;
> 			else
> 				resresv->svr_index = -1;
> 			if (resresv->svr_index == -1) {
> 				free_resource_resv(resresv);
> 				return NULL;
> 			}
> 		}
> ```
> 
> If I want to keep index as an integer itself I need to modify the following structure to contain one more field named numeric_value as follows. This new field is only used for the server_index attribute only. Hope that is fine.
> 
> ```
> struct attrl {
> 	struct attrl *next;
> 	char	     *name;
> 	char	     *resource;
> 	char	     *value;
>         unsigned    numeric_value:8,
> 	enum batch_op 	 op;	/* not used */
> };
> ```
> 
> Please let me know if this what you are suggesting.

No no, I am not suggesting to add anything to the attribute structure, rather to the job structure (inside scheduler) directly. On second thoughts, i now suggest, you leave it as is, since i will anyway change these to be a direct index based access

**Comment by suresh-thelkar** [comment]
> > > > I suggest we make the srv_index an byte (numeric) variable on the job structure itself (at the scheduler) that would avoid us needing to add yet another strcmp
> > > 
> > > 
> > > We already store it as a numeric on the job in scheduler
> > 
> > 
> > Yes we already store index as an integer(we can modify it as a bitfield of length 8) in both node_info and job_info structures of Scheduler. I think you mean to not convert the server_index to string in PBSD_status_get so that when we access it in the following places we can directly get it as an integer without again converting to integer using strtol().
> > ```
> > 
> > 		else if (!strcmp(attrp->name, ATTR_server_index)) {
> > 			count = strtol(attrp->value, &endp, 10);
> > 			if (*endp == '\0')
> > 				ninfo->svr_index = count;
> > 			else
> > 				ninfo->svr_index = -1;
> > 			if (ninfo->svr_index == -1) {
> > 				free_node_info(ninfo);
> > 				return NULL;
> > 			}
> > 		}
> > ```
> > 
> > 
> > and
> > ```
> > 		else if (!strcmp(attrp->name, ATTR_server_index)) {
> > 			count = strtol(attrp->value, &endp, 10);
> > 			if (*endp == '\0')
> > 				resresv->svr_index = count;
> > 			else
> > 				resresv->svr_index = -1;
> > 			if (resresv->svr_index == -1) {
> > 				free_resource_resv(resresv);
> > 				return NULL;
> > 			}
> > 		}
> > ```
> > 
> > 
> > If I want to keep index as an integer itself I need to modify the following structure to contain one more field named numeric_value as follows. This new field is only used for the server_index attribute only. Hope that is fine.
> > ```
> > struct attrl {
> > 	struct attrl *next;
> > 	char	     *name;
> > 	char	     *resource;
> > 	char	     *value;
> >         unsigned    numeric_value:8,
> > 	enum batch_op 	 op;	/* not used */
> > };
> > ```
> > 
> > 
> > Please let me know if this what you are suggesting.
> 
> No no, I am not suggesting to add anything to the attribute structure, rather to the job structure (inside scheduler) directly. On second thoughts, i now suggest, you leave it as is, since i will anyway change these to be a direct index based access

Thanks Subhasis. I am leaving it as is for now.
